packages/world-modules/test/DelegationModule.t.sol.sol

```diff
@@ -0,0 +1,106 @@
+
+import { DelegationModule } from "../src/modules/delegation/DelegationModule.sol";
+import { DelegationSystem } from "../src/modules/delegation/DelegationSystem.sol";
+import { getSignedMessageHash } from "../src/modules/delegation/getSignedMessageHash.sol";
+import { registerDelegationWithSignature } from "../src/modules/delegation/registerDelegationWithSignature.sol";
+
+contract DelegationModuleTest is Test, GasReporter {
+  using WorldResourceIdInstance for ResourceId;
+
+  IBaseWorld world;
+  DelegationModule delegationModule = new DelegationModule();
+
+  function setUp() public {
+    world = createWorld();
+    StoreSwitch.setStoreAddress(address(world));
+  }
+
+  function testInstallRoot() public {
+    startGasReport("install delegation module");
+    world.installRootModule(delegationModule, new bytes(0));
+    endGasReport();
+  }
+
+  function testRegisterDelegationWithSignature() public {
+    // Register a new system
+    WorldTestSystem system = new WorldTestSystem();
+    ResourceId systemId = WorldResourceIdLib.encode({
+      typeId: RESOURCE_SYSTEM,
+      namespace: "namespace",
+      name: "testSystem"
+    });
+    world.registerNamespace(systemId.getNamespaceId());
+    world.registerSystem(systemId, system, true);
+
+    world.installRootModule(delegationModule, new bytes(0));
+
+    // Register a limited delegation using signature
+    (address delegator, uint256 delegatorPk) = makeAddrAndKey("delegator");
+    address delegatee = address(2);
+
+    bytes32 hash = getSignedMessageHash(delegatee, UNLIMITED_DELEGATION, new bytes(0), 0, address(world));
+    (uint8 v, bytes32 r, bytes32 s) = vm.sign(delegatorPk, hash);
+    bytes memory signature = abi.encodePacked(r, s, v);
+
+    startGasReport("register an unlimited delegation with signature");
+    registerDelegationWithSignature(world, delegatee, UNLIMITED_DELEGATION, new bytes(0), delegator, signature);
+    endGasReport();
+
+    // Call a system from the delegatee on behalf of the delegator
+    vm.prank(delegatee);
+    bytes memory returnData = world.callFrom(delegator, systemId, abi.encodeCall(WorldTestSystem.msgSender, ()));
+    address returnedAddress = abi.decode(returnData, (address));
+
+    // Expect the system to have received the delegator's address
+    assertEq(returnedAddress, delegator);
+
+    // Unregister delegation
+    vm.prank(delegator);
+    world.unregisterDelegation(delegatee);
+
+    // Expect a revert when attempting to perform a call via callFrom after a delegation was unregistered
+    vm.expectRevert(abi.encodeWithSelector(IWorldErrors.World_DelegationNotFound.selector, delegator, delegatee));
+    vm.prank(delegatee);
+    world.callFrom(delegator, systemId, abi.encodeCall(WorldTestSystem.msgSender, ()));
+
+    // Attempt to register a limited delegation using an old signature
+    vm.expectRevert(abi.encodeWithSelector(DelegationSystem.InvalidSigner.selector, delegator, delegatee));
```

Comment:
I guess this error is more like `InvalidSignature`? (ie in this case the signer the same but the nonce is outdated, so the signature is not correct anymore?)

Reply:
We are now using `recover`, so I'm not sure what this outputs if the nonce is wrong 🤔 technically the recovered "signer" is incorrect either way